reps/2022-12-15-optimizer-data.md

```diff
@@ -0,0 +1,308 @@
+    * `ProjectionPushDownRule`: prune unused columns through plan as early as possible, up to datasource (ideally just read X columns from input, push SQL query into DB data source).
+
+### 3. Zero-copy vectorized execution with data batch
+
+#### 3.1. How things work today?
+
+* Each `Stage` in `ExecutionPlan` is executed sequentially. 
+* Each `Stage` materializes output as a list of `Block`s into the Ray Object Store, and this is served as input `Block`s for the next `Stage`. 
+* If multiple `OneToOneStage`s fused together into one `Stage`, each sub-stage still executed sequentially, with each sub-stage output `Iterable[Block]` in worker heap memory. The final sub-stage of the fused `Stage` materializes the `Block`s into the Ray Object Store. 
+* The read `Stage` can produce an `Iterable[Block]` with dynamic block splitting.
+
+Example:
+
+```
+Read() -> Map_batches(foo, batch_format="numpy", batch_size=N) -> Map_batches(bar, batch_format="numpy", batch_size=M)
+```
+
+Data flow:
+
+<img width="794" alt="data-flow-before" src="https://user-images.githubusercontent.com/4629931/207812215-db11f369-ef65-4e36-8070-0b4c4d46d2c9.png">
+
+Each in-memory Block is 512MB by default. The good thing here is we loosely follow the [Volcano iterator model](https://paperhub.s3.amazonaws.com/dace52a42c07f7f8348b08dc2b186061.pdf) + [vectorized execution](https://www.vldb.org/pvldb/vol11/p2209-kersten.pdf) (same design choice as [Databricks Photon](https://www.databricks.com/product/photon) and [Meta Velox](https://engineering.fb.com/2022/08/31/open-source/velox/)). The bad thing is to enforce `Block` size as 512MB, which may not work well with `batch_size` in `map_batches`. In `map_batches`, multiple `Block`s may be bundled or coalesced into one `Batch`, or one `Block` may be sliced into multiple `Batch`es. Also note, there's unnecessary data format conversion between NumPy batch and Arrow block. We should minimize the data copy and format conversion for intermediate in-memory data.
+
+#### 3.2. Proposed Change
+
+If users do not set `batch_size` in `map_batches`, we should choose a `batch_size `dynamically based on cluster resource, input size, estimated in-memory data size, etc. If users set `batch_size`, we respect the user provided `batch_size`.
+
+Use the `batch_size` from `map_batches`, to determine the `batch_size` in read stage. So each `Block` output by read is a batch directly passed to `map_batches` with zero-copy. The final output `Block`s to Object store can target 512MB though (may incur a copy for final output `Block`s).
+
+Data flow:
+
+<img width="795" alt="data-flow-after" src="https://user-images.githubusercontent.com/4629931/207812234-d94a6683-5e0d-4eb2-a7c3-6f7bb656f2b2.png">
+
+The new proposal would eliminate that batch to block copy and pass the batch directly to the following stage.
+
+#### 3.2.1. Interfaces
+
+NOTE: `OneToOneOperator` used here is the same as `OneToOneOperator` in "Native pipelining support in Ray Datasets" REP.
```

Comment:
This section needs to be updated, since the other REP now only proposes `PhysicalOperator`.

Reply:
@ericl - yeah updated. I need more thought to hook up `BatchedOperator.process_batches` with `PhysicalOperator.add_input/inputs_done/has_next/get_next`. But I think it should be implementation detail that we can figure it out later.

Reply:
@ericl - yeah updated. I need more thought to hook up `BatchedOperator.process_batches` with `PhysicalOperator.add_input/inputs_done/has_next/get_next`. But I think it should be implementation detail that we can figure it out later.